### PR TITLE
Right aligned backspace bug

### LINKED ISSUE #25199: TextField and TextFormField could not clear all the text in some situation
Hello, recently I encountered an problem, when I used the widget `TextField` and `TextFormField`, when I delete the characters which I typed and leave one character, then I found I could not delete the last character, below is the recurrent steps

## Steps to Reproduce

1. first let's new an TextField like this(below is the code)
```dart
new TextField(
        textAlign: TextAlign.right,
        style: new TextStyle(color: ColorCustom.common_item_content_normal, fontSize: 16.0,),
        controller: controller, 
        keyboardType: TextInputType.phone,
        decoration: new InputDecoration(
          border: InputBorder.none,
          icon: new Text("$title",style: new TextStyle(color: ColorCustom.color41, fontSize: 16.0),),
          hintStyle: new TextStyle(color: ColorCustom.common_item_content_no_edit, fontSize: 16.0),
          hintText: hint, 
        ),
      ),),
    );
```

2. above are an input box when running an app, then i type some number or other character,
below is  example:              1333333444444444484

3. and i put my cursor after 8, and before the last number 4(like this: 133333344444444448|4), then i began to type backspace to delete the number, until the last character  '4'  is remained

4. now i found i could not delete the character  '4' anymore

running `flutter doctor -v` 
```
C:\software\flutter\bin>flutter doctor -v
[√] Flutter (Channel stable, v1.0.0, on Microsoft Windows [Version 10.0.17134.407], locale zh-CN)
    • Flutter version 1.0.0 at C:\software\flutter
    • Framework revision 5391447fae (11 days ago), 2018-11-29 19:41:26 -0800
    • Engine revision 7375a0f414
    • Dart version 2.1.0 (build 2.1.0-dev.9.4 f9ebf21297)

[√] Android toolchain - develop for Android devices (Android SDK 28.0.3)
    • Android SDK at C:\Users\liuhuajie\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)
    • All Android licenses accepted.

[√] Android Studio (version 3.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 30.0.1
    • Dart plugin version 181.5656
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)

[!] IntelliJ IDEA Ultimate Edition (version 2017.2)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA 2017.2.3
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.io/intellij-setup/#installing-the-plugins

[√] Connected device (1 available)
    • PRO 6 • 80MBCNQ222N9 • android-arm64 • Android 7.1.1 (API 25)

! Doctor found issues in 1 category.
```

### PR DESCRIPTION
**Pull request by justinmc**
Closes https://github.com/flutter/flutter/issues/25199

In an input with `textAlign: TextAlign.right` with the cursor in the middle of the text:

```
    133333344444444448|4
```

Deleting until the end of the line caused the cursor to jump to the far right:

```
...
                    13|4
                     1|4
                       4|
```

This PR fixes it so the cursor stays in the correct place, so in this example it would be:

```
                      |4
```

I also encountered related bugs with multiline inputs while working on this.  When going to an empty new line with text align set to right, the cursor would jump back over to the far left.  These bugs seem to be fixed with this change as well.

### CODE REVIEW

**Comment by justinmc** [comment]
@HansMuller This is ready for a real review now.  Several tests were expecting incorrect behavior, so that's what all the lines changed in the test are.

File: packages/flutter/test/painting/text_painter_test.dart [hunk 1]
```diff
@@ -185,111 +188,339 @@ void main() {
...
+    caretOffset = painter.getOffsetForCaret(
+      ui.TextPosition(offset: offset, affinity: ui.TextAffinity.upstream),
+      ui.Rect.zero,
+    );
+    expect(caretOffset.dx, closeTo(0.0, 0.0001));
+    expect(caretOffset.dy, closeTo(SIZE_OF_A, 0.0001));
+
+    // When text wraps on its own, getOffsetForCaret disambiguates between the
+    // end of one line and start of next using affinity.
+    text = 'aaaaaaaa'; // Just enough to wrap one character down to second line
```
Comment (HansMuller) [inline packages/flutter/test/painting/text_painter_test.dart]: I assume that this wraps because SIZE_OF_A * 8 (112) is wider than the layout width of 100, specified below.  Maybe add a comment on the `painter.layout` line explaining what's up here.

File: packages/flutter/test/painting/text_painter_test.dart [hunk 2]
```diff
@@ -185,111 +188,339 @@ void main() {
...
+      ui.TextPosition(offset: offset, affinity: TextAffinity.upstream),
+      ui.Rect.zero,
+    );
+    expect(caretOffset.dx, closeTo(0.0, 0.0001));
+    expect(caretOffset.dy, closeTo(SIZE_OF_A, 0.0001));
 
-    // Correctly moves through second line with upstream affinity.
-    caretOffset = painter.getOffsetForCaret(const ui.TextPosition(offset: 4, affinity: TextAffinity.upstream), ui.Rect.zero);
-    expect(caretOffset.dx, closeTo(42.0, 0.0001));
+    // Check that getOffsetForCaret handles being at the end when affinity is
```
Comment (HansMuller) [inline packages/flutter/test/painting/text_painter_test.dart]: I think it would be better to explain what the test is checking, like the previous test does (and the next one as well). Something like:  Given a one-line right aligned string, positioning the cursor at offset 0 means that it appears at the "end" of the string, at x = 0.

File: packages/flutter/test/painting/text_painter_test.dart [hunk 3]
```diff
@@ -185,111 +188,339 @@ void main() {
...
+    expect(caretOffset.dy, closeTo(SIZE_OF_A, 0.0001));
+    caretOffset = painter.getOffsetForCaret(
+      ui.TextPosition(offset: offset, affinity: TextAffinity.upstream),
+      ui.Rect.zero,
+    );
+    expect(caretOffset.dx, closeTo(0.0, 0.0001));
+    expect(caretOffset.dy, closeTo(SIZE_OF_A, 0.0001));
 
-    // Correctly handles multiple trailing newlines.
+    // Correctly handles multiple trailing newlines with the same offset
```
Comment (HansMuller) [inline packages/flutter/test/painting/text_painter_test.dart]: As before, best to explain what we expect here.

**Review by HansMuller — APPROVED** [review]
As far as I can tell, this all looks correct. LGTM

It would be helpful to add some additional explanation to the TextAffinity enum doc, particularly about white space, implicit newlines ("forced line breaks"), genuine newlines, and bidi text. Many of the corner cases covered by the tests probably deserve  a sentence or two. Given that, other places that would benefit from the clarification, like the TextPosition class, could remind readers to "See Also" the TextAfinity class for more information about cursor positioning...

Updating the TextAffinity et al docs could be a separate PR.

**Comment by justinmc** [comment]
@HansMuller I just improved the comments and cleaned the tests up a bit.  I'll open a PR in the engine to improve the TextAffinity docs and reference it here.

**Comment by justinmc** [comment]
@GaryQian Heads up that I'm changing similar code in text_painter.dart to what you just changed.  I just merged your changes into this PR and I think it went well though.  I think you and I were partially fixing the same bug in `_getOffsetFromDownstream`.  I really just added [this check for an edge case](https://github.com/flutter/flutter/pull/25229/files#diff-f33d2b321b2d94f3ad59fb667ba846ccR455) and cleaned up the tests a bit.  They seemed to be mistaking newline characters and wrapping, which has different behavior with TextAffinity.

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 4]
```diff
@@ -446,13 +449,22 @@ class TextPainter {
...
+      }
+
       final double caretEnd = box.end;
       final double dx = box.direction == TextDirection.rtl ? caretEnd - caretPrototype.width : caretEnd;
       return Offset(dx, box.top);
     }
     return null;
   }
 
+  // Get the Offset of the cursor (in pixels) based off the near edge of the
```
Comment (GaryQian) [inline packages/flutter/lib/src/painting/text_painter.dart]: May want to distinguish between logical vs physical pixels, as 'pixels' may be ambiguous. In this case I think everything should be logical pixels.

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 5]
```diff
@@ -417,12 +417,15 @@ class TextPainter {
   // Unicode value for a zero width joiner character.
   static const int _zwjUtf16 = 0x200d;
 
+  // Get the Offset of the cursor (in pixels) based off the near edge of the
```
Comment (GaryQian) [inline packages/flutter/lib/src/painting/text_painter.dart]: See below

**Comment by justinmc** [comment]
@GaryQian I clarified "logical pixels" in the docs like you mentioned, but I couldn't find a better way to check for newlines other than the integer value 10.  It looks like `\r` is also a typical newline value for example, but it is displayed as a space in our text fields and not as a newline, so we shouldn't match it here.  Let me know if you can think of anything else though.

**Review by GaryQian — APPROVED** [review]
